pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -96,63 +109,132 @@ func (e *DockerRegistryServiceController) Stop() {
+			net.JoinHostPort(fmt.Sprintf("%s.%s.svc.cluster.local", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port))}
 	}
 
-	return e.defaultDockerURL
+	return []string{e.defaultDockerURL}
 }
 
 // serviceAdded reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
 // changing all interestedURLs
 func (e *DockerRegistryServiceController) serviceAdded(obj interface{}) {
-	service := obj.(*api.Service)
-	if service.Name != e.registryServiceName {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
 		return
 	}
 
-	if err := e.handleLocationChange(e.getServiceLocation(service)); err != nil {
-		util.HandleError(err)
+	service := obj.(*api.Service)
+	if service.Name == e.registryServiceName {
+		e.markForChange()
 	}
 }
 
 // serviceUpdated reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
 // changing all interestedURLs, if needed
 func (e *DockerRegistryServiceController) serviceUpdated(oldObj interface{}, newObj interface{}) {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
+		return
+	}
+
 	oldService := oldObj.(*api.Service)
 	newService := newObj.(*api.Service)
 	if newService.Name != e.registryServiceName {
 		return
 	}
-	if e.getServiceLocation(oldService) == e.getServiceLocation(newService) {
-		return
-	}
 
-	if err := e.handleLocationChange(e.getServiceLocation(newService)); err != nil {
-		util.HandleError(err)
+	oldServiceKeys := e.getServiceLocations(oldService)
+	newServiceKeys := e.getServiceLocations(newService)
+	if !reflect.DeepEqual(oldServiceKeys, newServiceKeys) {
+		e.markForChange()
 	}
+
 }
 
 // serviceDeleted reacts to the docker-registry deletion by updating all the generated dockercfg secrets
 func (e *DockerRegistryServiceController) serviceDeleted(obj interface{}) {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
+		return
+	}
+
 	service, ok := obj.(*api.Service)
 	if !ok {
 		return
 	}
-	if service.Name != e.registryServiceName {
-		return
+	if service.Name == e.registryServiceName {
+		e.markForChange()
 	}
 
-	if err := e.handleLocationChange(e.defaultDockerURL); err != nil {
-		util.HandleError(err)
-	}
 }
 
 // handleLocationChange goes through all service account dockercfg secrets and updates them to point at a new docker-registry location
-func (e *DockerRegistryServiceController) handleLocationChange(serviceLocation string) error {
-	e.dockercfgController.SetDockerURL(serviceLocation)
+func (e *DockerRegistryServiceController) handleLocationChange() error {
+	service, serviceExists, err := e.serviceCache.Get(&api.Service{ObjectMeta: api.ObjectMeta{Namespace: e.registryNamespace, Name: e.registryServiceName}})
```

Comment:
This causes me an immense amount of physical pain to see.  This seems like an enormous amount of tricky controller code to support something that should never change, or be an admin fix up command.

If we used DNS for internal, the first name would be stable.  If we (in DNS) supported cname to the route, that would never change.  Even without that, why would I just run a fix command for this.